### ISSUE TITLE
investigate better approach to patchedast

### ISSUE THREAD

**Issue by enomado** [python-rope/rope#237]
Just to summarize some possible alternatives:

- [RedBaron](https://github.com/PyCQA/redbaron)
- [astroid](https://github.com/PyCQA/astroid)
- [parso](https://github.com/davidhalter/parso)

**Comment by soupytwist** [comment]
I'm familiar with these. What's the problem you are trying to resolve that makes any of these better?

**Comment by enomado** [comment]
Current architecture of patchedast is limited and can't parse all code. It can be proved by traversing site-packages and applying it to files. Or just write some rich function call or multiline string with nested bracket to feel that the task requires full parsing approach instead of ad-hoc lookahead and manual assumptions about a next node.

**Comment by emacsway** [comment]
> Current architecture of patchedast is limited and can't parse all code.

How patchedast is related to parsing? Do you understand what parsing is about? Rope uses python's native parser https://docs.python.org/3/library/parser.html

> architecture of patchedast is limited

What exactly limitation on architecture level do you mean?

I understand some advantages of astroid, but they mainly are not related to parsing.

**Comment by enomado** [comment]
> How patchedast is related to parsing? Do you understand what parsing is about? Rope uses python's native parser
Well okay. What do you think patchedast do ? It `parses` python code with little help of parsed ast because python's parser (surprisingly) cant parse pythons code with column resolution.

> What exactly limitation on architecture level do you mean?
I mean that its exactly not a parser, but should be. Currently it tries to assume `pos` of nodes ad-hoc, without any 'parsing things' like stacks, automata (sorry for such silly terms, I guess you got what I mean).

**Comment by emacsway** [comment]
> It parses python code with little help of parsed ast because python's parser (surprisingly) cant parse pythons code with column resolution.

- Can you give me any test case to reproduce any problem of python's native parser?

**Comment by enomado** [comment]
Existence of https://github.com/python-rope/rope/blob/master/rope/refactor/patchedast.py is a good proof I guess. In other case I'll very happy.

**Comment by emacsway** [comment]
At what line of the file do you see a good proof?

**Comment by enomado** [comment]
Whole file, purpose of the file. Stop trolling please. If you are sure the problem is 100% gone since 2012 and python parser have good column resolution so okay then we can close the ticket...

**Comment by emacsway** [comment]
Can you call me with Skype to discuss this? My skypename is industrialnet.

**Comment by emacsway** [comment]
yes, astroid solves some issues of rope, but its implementation requires a lot of works.

**Comment by bageljrkhanofemus** [comment]
I've messed around with removing the patchedast. It provides two things:
1. Sorted children, which afaik isn't directly used
2. Region of code.

Since rope reads everything into a single line, using the standard ast would break region based functionality. Regardless, I am working on a solution using standard ast.  If we can use parso, we'll get error recovery and port some of my autoimport code from pylsp.
